### DISCUSSION TITLE
How to insert nodes with proper position data?

### DISCUSSION THREAD

**Discussion by AshtarCodes** [rehypejs/rehype#84] [Q&A] [answered]
I have developed a sort of codemod /  syntax conversion tool using rehype, and I am inserting nodes at various points of the AST, some are elements, most are comments and new lines. I would like for users of the tool to clearly see what changes the tool has done. Unfortunately, my file loses its previous formatting / spacing when I insert nodes, which then makes file diffs unreadable. I understand that this happens because I am not providing accurate `position` data when I insert nodes. I am wondering, is there a best practice way of determining and updating `position` data of all nodes in the tree?  

I've looked at the tree traversal guides and I don't see any mention of dealing with `position`. I do see that `unist-util-position` lets me get the position data of a given node. Maybe there is a suggested way of using this util? Thanks in advance!

**Comment by wooorm** [comment, marked as answer]
Your problem most likely does not have to do with `position`s. Your problem has to do with ASTs: ASTs are abstract. They don’t have information about which quote was used (`"` or `'`) and other “insignificant” information.
When you generate output from an AST, it will often look different from the input with white-space and the like.

You can either:
- first reformat (parse and serialize) and save that result as X, then perform your codemod on X and save the result as Y, and finally compare X and Y. 
- Diff in whitespace-insensitive mode
- Probably much more, but I’d probably need to learn more about your setup first

**Reply by AshtarCodes** [reply]
Okay, well thanks for saving me the time of messing around with `position`s. I think I'll ask the users of the tool to format their files with something like Prettier before and after using the tool. That will also help clean up the diff. Thanks for taking the time to answer @wooorm
